### PR TITLE
feat: add 5-band EQ equalizer with auto preamp and soft limiter

### LINKED ISSUE #86: feat: add 5-band EQ equalizer with auto preamp and threshold-based soft limiter
# feat: add 5-band EQ equalizer with auto preamp and threshold-based soft limiter

> One issue = one PR. Implement after #85 is merged. This plan is authoritative for the post-#85 coordinator architecture.

## Outcome

Add a five-band peaking EQ at 60 Hz, 230 Hz, 910 Hz, 3.6 kHz and 14 kHz, each with Q = 0.707 and gain range `-12.0..=12.0 dB`. Add click-free enable/gain transitions, conservative automatic preamp, a bounded soft limiter, persisted settings and an accessible Settings UI.

The final render order is fixed:

```text
existing source/stem mix + master/stem gains
→ EQ dry/wet processor + auto preamp
→ soft limiter
→ existing play/pause/seek fade
→ output/AirPlay forwarding
```

Issue #87 later samples the signal after the fade. Preserve that insertion point.

## Non-goals

- FFT, spectrum analysis, per-stem EQ, presets or user-defined band frequencies
- compressor makeup gain, loudness normalization or look-ahead limiting
- changing existing playback snapshot fields or volume semantics
- heap allocation, blocking, logging or coefficient error propagation from the CPAL callback

## Dependency and API decisions

Add:

```toml
biquad = "0.5"
```

Use the crate's actual 0.5 API:

```rust
Coefficients::<f32>::from_params(
    Type::PeakingEQ(gain_db),
    sample_rate.hz(),
    frequency.hz(),
    EQ_Q,
)
```

`from_params` returns `Result`; never call `unwrap`/`expect` in the realtime path. `DirectForm1::update_coefficients` preserves filter state and performs no allocation.

### Rust config/settings

Add to `AppConfig`:

```rust
#[serde(skip_serializing_if = "Option::is_none")]
pub eq_enabled: Option<bool>,
#[serde(skip_serializing_if = "Option::is_none")]
pub eq_gains_db: Option<[f32; 5]>,
```

Effective defaults are `false` and `[0.0; 5]`. Add both resolved fields to `AppSettings`:

```rust
pub eq_enabled: bool,
pub eq_gains_db: [f32; 5],
```

Add settings commands returning a complete `AppSettings` snapshot:

```rust
set_eq_enabled(enabled: bool) -> AppSettings
set_eq_gains(gains_db: [f32; 5]) -> AppSettings
```

Validation is exact:

- array length is enforced by the Tauri/Rust type
- every value must be finite
- every value must be in `-12.0..=12.0`; reject the whole request instead of clamping
- error mapping is the existing `invalid_playback_state` code with `keep_current_state`; do not expand the shared error enum in this PR

Do not add a separate `get_eq_state`; startup `get_settings` is the single authority.

### Coordinator integration

Extend #85's `PlaybackCommand` only in this PR:

```rust
SetEqEnabled { enabled: bool, reply: EqReply },
SetEqGains { gains_db: [f32; 5], reply: EqReply },
```

The coordinator updates the controller configuration and returns `EqState { enabled, gains_db }`. Settings commands use this sequence:

1. validate input;
2. read the previous persisted/runtime values;
3. send the coordinator update and await acknowledgement;
4. persist the complete config;
5. if persistence fails, send a best-effort coordinator rollback to the previous values, log rollback failure and return the persistence error;
6. return `settings_from_config` on success.

Initialize the controller from effective config before spawning the coordinator/output thread in `setup_app()`. Tests and `PlaybackController::default()` remain EQ-off/flat.

Update `docs/references/contracts/playback.md`, `src/types/ipc.ts`, Tauri wrappers, IPC registration tests and both locale files.

## DSP implementation

Create `src-tauri/src/audio/eq.rs`.

```rust
pub const EQ_BAND_FREQUENCIES_HZ: [f32; 5] =
    [60.0, 230.0, 910.0, 3_600.0, 14_000.0];
pub const EQ_Q: f32 = 0.707;
pub const EQ_MIN_GAIN_DB: f32 = -12.0;
pub const EQ_MAX_GAIN_DB: f32 = 12.0;
const EQ_SMOOTH_MS: f32 = 50.0;
const BYPASS_SMOOTH_MS: f32 = 20.0;
const NYQUIST_RATIO_LIMIT: f32 = 0.45;
const LIMITER_THRESHOLD: f32 = 0.95;
```

`EqProcessor` is owned by the CPAL output closure beside `ResamplerCache`; it is not stored behind the playback mutex. Construct it once from the actual device sample rate/channel count. A new stream constructs a new processor.

Required state:

```rust
pub struct EqProcessor {
    filters: Vec<[Option<DirectForm1<f32>>; 5]>, // channel → band
    sample_rate: f32,
    channels: usize,
    enabled: bool,
    target_gains_db: [f32; 5],
    current_gains_db: [f32; 5],
    target_preamp: f32,
    current_preamp: f32,
    wet_mix: f32,
    target_wet_mix: f32,
}
```

The precise algorithm:

1. `PlaybackController` stores private `eq_enabled`, `eq_gains_db` and monotonically increasing `eq_revision`; expose validated setters and a copied config snapshot. Do not expose mutable public fields.
2. The callback compares `eq_revision` with its last applied revision while it already holds the controller lock, then calls `set_enabled`/`set_gains` on its local processor.
3. Gain, preamp and dry/wet transitions advance by rendered frames, derived from the device sample rate. They do not advance on zero-length/buffering callbacks.
4. Recompute coefficients at most once per callback per band using the current smoothed gain. Apply them with `update_coefficients` so delay state remains warm.
5. For `frequency >= sample_rate * 0.45`, keep that channel/band as `None`. A coefficient error also disables only that band for that callback; retain the last valid coefficient and expose no panic/log storm.
6. Process channels independently over interleaved frames. For each sample retain one scalar `dry`, run the wet filter chain, apply smoothed preamp to wet, then mix `dry * (1 - wet_mix) + wet * wet_mix` in place. No temporary audio buffer is allocated.
7. Continue running valid filters while bypassed so re-enable uses warm state. Do not add a flat/bypassed fast path in this PR.
8. Auto preamp target is `db_to_linear(-max(0, max_positive_gain))`; smooth it over the same 50 ms interval. It supplies headroom and does not replace the limiter.
9. Process only `&mut output[..rendered_samples]`; trailing callback padding must not advance filter state.

### Correct soft limiter

The previous issue formula used a threshold of `1.0`, which made `(1.0 - threshold)` zero and reduced the branch to a hard clip. Use this continuous, slope-matched curve:

```rust
fn soft_limit(sample: f32) -> f32 {
    if !sample.is_finite() {
        return 0.0;
    }
    let magnitude = sample.abs();
    if magnitude <= LIMITER_THRESHOLD {
        return sample;
    }
    let headroom = 1.0 - LIMITER_THRESHOLD;
    let compressed = LIMITER_THRESHOLD
        + headroom * ((magnitude - LIMITER_THRESHOLD) / headroom).tanh();
    sample.signum() * compressed
}
```

This is continuous with derivative 1 at the threshold, leaves samples at or below `0.95` bit-for-bit unchanged, remains finite and asymptotically stays below `1.0`. Replace all three existing per-branch hard-clamp loops with one limiter pass after EQ and before fade.

## Output wiring

Update `build_output_stream` to create the processor before moving into the callback, and extend `render_output_buffer` with `&mut EqProcessor`. Keep all existing streaming/stem/fallback return counts and frame advancement intact.

No callback operation may:

- lock a second mutex
- allocate after steady-state initialization
- build a `Vec`/`String`
- log per sample/callback
- serialize or emit an event

## Settings UI

Create `src/components/Settings/SettingsEqSection.tsx` and render it after General settings.

- top row: native checkbox/switch labelled Equalizer
- five rows: `60 Hz`, `230 Hz`, `910 Hz`, `3.6 kHz`, `14 kHz`
- each range: min `-12`, max `12`, step `0.5`; visible signed dB value; disabled while EQ is off
- Reset button restores all bands to `0.0` and persists once
- local slider draft updates immediately; use one 75 ms trailing call carrying the complete five-value array, flush on pointer/key release and cancel on unmount
- API failure restores the latest authoritative store values and uses `notifyError`
- labels, reset text and descriptions use `en.json` and `zh-CN.json`
- keyboard arrows, Home/End and focus-visible behavior remain native

Extend settings-store snapshot/sync serialization with `eqEnabled`, `eqGainsDb`, `setEqEnabled` and `setEqGains`. Do not use localStorage.

## Files

Create:

- `src-tauri/src/audio/eq.rs`
- `src/components/Settings/SettingsEqSection.tsx`
- focused Rust and frontend tests

Modify:

- `src-tauri/Cargo.toml`, `Cargo.lock`
- `src-tauri/src/audio/mod.rs`, `audio/output.rs`, `audio/playback.rs`, `audio/coordinator.rs`
- `src-tauri/src/config.rs`, `commands/settings.rs`, `lib.rs`
- `docs/references/contracts/playback.md`
- `src/types/ipc.ts`, `src/lib/tauri/settings.ts`, barrel/wrapper/contract tests
- `src/stores/settings-store.ts` and tests
- `src/components/Settings/SettingsOverlay.tsx` and tests
- `src/locales/en.json`, `src/locales/zh-CN.json`
- `CHANGELOG.md`

## Required tests

Rust tests must cover:

- flat/bypassed path error `<= 1e-6` for values below limiter threshold
- exact limiter continuity around `0.95`, bounded output and non-finite sanitization
- positive/negative symmetry
- +12 dB band response versus an off-band tone
- auto-preamp target and smoothed transition
- bypass and gain changes with a bounded adjacent-sample discontinuity
- independent left/right filter state
- 22.05/32/44.1/48/96 kHz Nyquist guards
- invalid gains rejected without changing state/config
- config hydration and persistence rollback
- existing render tests for streaming, two-stem, four-stem, fade and AirPlay

Frontend tests must cover full settings hydration/sync, debounce/flush/cancel, reset, disabled controls, optimistic draft rollback, i18n and Tauri wrapper arguments.

## Verification

```bash
pnpm format
pnpm lint
pnpm check:i18n
pnpm knip --no-progress
pnpm build
pnpm test
pnpm test:coverage
pnpm check:patch-coverage
cd src-tauri && cargo fmt --check && cargo test -q
cd .. && pnpm tauri build
```

Run `cargo deny check` because this PR adds a dependency. The PR must include exact results, a short realtime-allocation review, and manual listening checks at flat, boost, cut, rapid bypass and 22.05/48 kHz output.

## Acceptance checklist

- [ ] Settings persist and hydrate before audio startup.
- [ ] All gains are validated and coordinator-serialized.
- [ ] The processor lives in the output closure and preserves per-channel state.
- [ ] Flat and bypassed audio is transparent below `0.95`.
- [ ] Limiter is continuous and bounded; the old hard clamps are gone.
- [ ] No steady-state callback allocation, blocking or event emission is introduced.
- [ ] UI is accessible, localized and rolls back on persistence failure.
- [ ] Full media-path tests and packaged build pass.

## Dependency order

Blocked by #85. Blocks #87 and #88 so later output-path PRs preserve this exact signal chain.

### PR DESCRIPTION
**Pull request by thedavidweng**
## Summary

- Add a 5-band peaking EQ (60 Hz, 230 Hz, 910 Hz, 3.6 kHz, 14 kHz; ±12 dB per band) with automatic preamp reduction to compensate for positive boosts and a tanh-based soft limiter that engages above 0.85 amplitude
- Wire `EqProcessor` into the realtime output callback after the source/stem mix and before the play/pause fade, with linear smoothing for gain, preamp, and bypass dry/wet transitions
- Add `set_eq_enabled` and `set_eq_gains` Tauri commands that persist to `AppConfig` and push updates to the active `PlaybackController` via `PlaybackCoordinator`
- Add a new `SettingsEqSection` in the settings overlay with per-band sliders, enable toggle, and reset-to-flat button
- Add `eq_enabled` and `eq_gains_db` fields to `AppSettings` IPC contract
- Update i18n strings (en, zh-CN), CHANGELOG, and playback contract docs
- Update Flatpak `cargo-sources.json` with `biquad` and `libm` dependencies

Closes #86

#### Test plan

- [x] 417 Rust tests pass (`cargo test --lib`)
- [x] 1323 frontend tests pass (`pnpm vitest run`)
- [x] `cargo clippy --lib` — no warnings
- [x] `pnpm tsc --noEmit` — no type errors
- [x] `pnpm tauri build --debug --no-bundle` — builds successfully
- [x] Patch coverage 98.1% (target 80%)
- [x] All pre-push hooks pass (lint, format-check, cargo-fmt-check, patch-coverage)

---

<a href="https://app.devin.ai/review/thedavidweng/openkara/pull/114" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

### CODE REVIEW

**Comment by thedavidweng** [comment]
## Maintainer acceptance — merge slot 6 (audio-chain foundation)

**BLOCKED — changes required.**

### Blocking DSP correctness issue

`EqProcessor::process` computes each gain/preamp/wet step once per callback as `(target - current) / smooth_samples`, then recomputes a smaller step on the next callback. With normal 256/512-frame CPAL callbacks this is an exponential, buffer-size-dependent transition, not the specified 50 ms linear transition. The existing 50 ms test passes only because it processes all 2,400 frames in one large call.

For example, at 48 kHz with 256-frame callbacks, after 2,400 rendered frames a substantial fraction of the original delta remains. This makes smoothing duration depend on callback chunking and fails #86's click-free fixed-duration contract.

Store remaining transition frames/slope when a target changes (or use an equivalent chunk-invariant formulation), and add tests that feed the same 50 ms change through arbitrary callback chunk boundaries. Apply the same correction to gains, preamp, and wet/bypass interpolation.

The older rollback/stale-draft threads appear fixed in the current head; this finding is independent and blocks the audio foundation.

File: src/components/Settings/settings-overlay.library-actions.ts [hunk 1]
```diff
@@ -324,5 +327,52 @@ export function createLibrarySettingsActions(
...
+      next[band] = clamped;
+      patchState({ eqGainsDb: next });
+      dependencies.settingsStore.patchAppSettings({ eqGainsDb: next });
+
+      try {
+        const settings = await dependencies.api.setEqGains(next);
+        dependencies.settingsStore.hydrateAppSettings(settings);
+        patchState({ eqGainsDb: settings.eq_gains_db });
+      } catch (error) {
+        dependencies.notifyError(error);
```
[the remark this answers is not in the document]
Comment (thedavidweng) [reply src/components/Settings/settings-overlay.library-actions.ts:371]: Fixed on HEAD: `setEqEnabled`, `setEqGains`, `setEqBandGain`, and `resetEqGains` capture previous values before optimistic patch and restore both overlay state and the settings store in `catch`. See `settings-overlay.library-actions.ts`.

File: src-tauri/src/commands/settings.rs [hunk 2]
```diff
@@ -193,6 +197,78 @@ pub fn set_execution_provider(
...
+    let (tx, rx) = tokio::sync::oneshot::channel();
+    let command = crate::audio::coordinator::PlaybackCommand::SetEqEnabled { enabled, reply: tx };
+    state
+        .playback
+        .command_tx
+        .send(command)
+        .map_err(|_| internal_error("playback coordinator disconnected"))?;
+    let _ = rx
+        .blocking_recv()
+        .map_err(|_| internal_error("playback coordinator dropped reply"))?;
```
[the remark this answers is not in the document]
Comment (thedavidweng) [reply src-tauri/src/commands/settings.rs]: Fixed: `set_eq_enabled` / `set_eq_gains` now map the coordinator `Result` through `?` so `PlaybackError` is returned to the caller instead of being discarded with `let _`.

**Comment by thedavidweng** [comment]
## Re-review: blocking DSP correctness fix required

This PR is **not ready to merge**. The latest head still recomputes the smoothing step on every audio callback, e.g. `step = (target - current) / smooth_samples`. That produces exponential convergence whose timing depends on callback chunking; it is not the specified 50 ms linear transition. A single 2,400-frame test callback masks the defect.

Implement exactly the following:

1. Move transition scheduling out of the per-callback calculation. When EQ gains, preamp, or enabled/bypass state changes, capture the current values, calculate the target values once, and initialise a persistent transition state with:
   - a remaining sample-frame count of `round(sample_rate * 0.050)`;
   - fixed per-sample deltas (or an equivalent remaining-frames interpolation) for every band gain, preamp, and wet/dry value.
2. In `EqProcessor::process`, consume that persistent state sample by sample. A callback may process any number of frames; it must only advance the transition by exactly that many frames. Do **not** recalculate a smaller denominator on the next callback.
3. If settings change during an active ramp, start a new 50 ms ramp from the actually current interpolated values. Preserve the realtime-callback allocation-free behavior and existing audio ordering.
4. Cover gains, auto-preamp, and enabled/bypass wet-dry—not just one gain path.

Add regression tests that compare the same 50 ms transition rendered:
- in one callback;
- across deliberately uneven callback partitions (for example 9, 127, 256, 1,024 frames);
- at 48 kHz and a non-48 kHz rate.

The output/parameter values must agree at equal sample positions regardless of partitioning, and must reach the target with no residual delta at the exact transition boundary. Keep the existing finite-value and limiter assertions.

Before requesting re-review, rebase onto the then-current `main` (after prerequisite merges), run the existing Rust/frontend suites, and report the new commit SHA plus the exact new test names/results.

**Comment by thedavidweng** [comment]
## Re-review — still blocked: EQ audio output remains callback-partition dependent

Reviewed at `850cf0c`. The scalar ramp state is now persistent, but the audible DSP path is still not chunk-invariant:

- `EqProcessor::process` refreshes every biquad's coefficients **once before** the per-frame smoothing loop, from `current_gains_db`.
- It then advances `current_gains_db` per frame while processing samples with those old, callback-entry coefficients.

Consequently, a single 2,400-frame callback at 48 kHz renders the entire 50 ms interval with the pre-change filter coefficients; a 256-frame partition refreshes at each callback boundary. The resulting audio differs by CPAL callback partition even though the scalar reaches its target.

Make the filter response follow the same deterministic per-frame schedule:

1. Keep the fixed 50 ms scalar ramps, but move coefficient application into the frame schedule. For each rendered frame, advance the scalars, derive the five coefficients once for that frame, update the corresponding per-channel filters, then process that frame. Do not defer coefficient changes to the next callback.
2. Keep this allocation-free in the realtime callback and preserve each `DirectForm1` delay state; only coefficients should change.
3. Add a regression rendering the same tone/impulse through one 2,400-frame call and uneven partitions (for example 9, 127, 256, 1,024, remainder), at both 48 kHz and 44.1 kHz. Compare **audio output** and parameter state at equal sample positions within a justified float tolerance.
4. Assert that gain, auto-preamp, and wet/bypass all reach their targets exactly at the requested frame boundary, with no residual callback-sized delay.

The existing 256-frame scalar-only assertion is useful but does not cover this audible defect.

**Comment by thedavidweng** [comment]
## Re-review — blocked: EQ settings synchronously wait on the UI command path

Reviewed at `6453023`. Both `set_eq_enabled` and `set_eq_gains` are synchronous Tauri commands and call `rx.blocking_recv()` after enqueueing the coordinator command. A coordinator round-trip can sit behind load/output work, so a slider drag or toggle can block the window event path for seconds.

Please remove the synchronous wait from these commands:

1. Make the command handlers asynchronous and await the reply, or move the complete blocking transaction to `tauri::async_runtime::spawn_blocking`; do not use `blocking_recv()` on the window command path.
2. Preserve the current validation and coordinator-error propagation. Keep persisted settings and the audio engine semantically consistent on failure rather than reporting an error after leaving a divergent state.
3. Add focused coverage for the delayed-coordinator path (and retain the existing success/error tests) so a future refactor cannot reintroduce a blocking wait.

Separately, the latest required frontend-test step exits non-zero and emits no `coverage/coverage-summary.json`; the downstream coverage reporter consequently fails as well. Please make the exact CI command green before requesting another review:

```
node --run test -- --coverage --reporter=junit --outputFile=test-results.junit.xml
```

No LGTM while either the UI-thread wait or the red CI remains.

**Comment by thedavidweng** [comment]
## Re-review — still blocked: coordinator failures leave persisted EQ settings divergent from the running engine

Reviewed at `b8cf7ac`. Converting the commands to `async fn` and awaiting the oneshot removes the UI-thread wait — that part is correct.

But both handlers still do this order:

```
load config → mutate config → save config → send coordinator command → await result
```

If the channel is disconnected, the reply is dropped, or the coordinator rejects the update, the command returns an error **after the new EQ value is already persisted**. The audio engine remains on the old value until restart, violating the required runtime/persistence consistency.

Please make the operation failure-atomic:

1. Capture the full old config value (including `Option` state).
2. Apply the change through the coordinator first; do not persist on any send/reply/apply failure.
3. Persist only after the coordinator succeeds. If persistence then fails, send and await the inverse coordinator command so the running engine returns to the old value before reporting the error.
4. Do not hold a mutex/guard across an `.await`.

Add focused tests for coordinator-send/reply/apply failure and persistence failure, proving both the stored config and active controller stay at the previous values. No LGTM yet.

**Comment by thedavidweng** [comment]
@greptile review

**Comment by thedavidweng** [comment]
@greptile review

**Comment by thedavidweng** [comment]
@greptile review

**Comment by thedavidweng** [comment]
@greptile review

**Comment by thedavidweng** [comment]
Code re-review of head `6c616cb9ab713fdebcc81a44ee0020c40d4f694a` passed.

The async coordinator wait remains non-blocking, persistence rollback is preserved, the toggle now restores the captured previous authoritative value on failure, and reset cancels any pending debounced slider commit before writing the flat gains. CI and Packaging are green on this exact head.

No additional automated code blocker found. Final LGTM remains gated on the declared manual EQ UI and audible DSP/bypass/device-restart acceptance.

File: src-tauri/src/audio/output.rs [hunk 3]
```diff
@@ -358,36 +353,38 @@ pub fn render_output_buffer(
...
+    //   → output/AirPlay forwarding
+    // EQ smoothing advances only on rendered samples; trailing callback
+    // padding (zero-filled above) must not advance filter state.
+    eq_processor.process(output, rendered);
+
+    // Soft limiter: bound peaks to prevent clipping after EQ boost. Samples
+    // below the threshold pass through unchanged.
+    for sample in output[..rendered].iter_mut() {
+        *sample = soft_limit(*sample);
+    }
```
[the remark this answers is not in the document]
Comment (thedavidweng) [reply src-tauri/src/audio/output.rs]: Fixed in commit 1ea15a6. Added `EqProcessor::is_fully_bypassed()` and gated the soft limiter in `render_output_buffer` so it only runs when the EQ is active or transitioning (wet mix non-zero or ramping). For the default no-EQ single-source path the limiter is now skipped entirely, preserving bit-transparent playback for in-range audio. The limiter still engages when EQ boosts or stem summing could push levels past 1.0. Added test `is_fully_bypassed_true_by_default_and_after_disable_settles`.

[the remark this answers is not in the document]
Comment (thedavidweng) [reply src-tauri/src/audio/output.rs]: test

File: src/components/Settings/SettingsEqSection.tsx [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+                disabled={meta.isInitializing || !state.eqEnabled}
+                className="w-full accent-[var(--color-accent)]"
+              />
+            </div>
+          ))}
+
+          <button
+            type="button"
+            onClick={() => void actions.resetEqGains()}
+            disabled={meta.isInitializing || !state.eqEnabled}
```
[the remark this answers is not in the document]
Comment (thedavidweng) [reply src/components/Settings/SettingsEqSection.tsx:164]: Fixed in commit 9e29433. resetEqGains now clears pendingRef and timerRef before calling actions.resetEqGains(), so a pending debounce cannot fire afterwards and restore pre-reset gains.

File: src/stores/settings-store.ts [hunk 5]
```diff
@@ -158,6 +177,44 @@ export function createSettingsStore(
...
+        // Optimistically update local state so the toggle reflects immediately.
+        syncPatch({ eqEnabled: enabled });
+        try {
+          const settings = await api.setEqEnabled(enabled);
+          syncPatch(toAppSettingsSnapshot(settings));
+        } catch (error) {
+          // Revert on failure.
+          syncPatch({ eqEnabled: !enabled });
+          notifyError(error);
+        }
```
[the remark this answers is not in the document]
Comment (thedavidweng) [reply src/stores/settings-store.ts:213]: Fixed in commit 9e29433. setEqEnabled now captures const previous = get().eqEnabled before the optimistic patch and reverts to previous (not !enabled) on failure.

**Comment by thedavidweng** [comment]
## Re-review — fixes are correct, but this branch must be rebuilt on current main

Reviewed head `3fb8c6227f2b1ef151605ef8afb7e6ff2b71c74a`.

The latest head correctly adds a shared EQ mutation generation (discarding stale toggle/gain responses) and restores hard clipping protection when EQ is fully bypassed. CI and Packaging are green on this head.

No LGTM yet because #103 has already merged an **older EQ implementation** into `main`, and this PR now conflicts. Rebase onto current `main` and reduce the PR to the fixes missing from `main`; do not replace or lose the merged gapless/PeakMeter paths while resolving `output.rs`.

Required post-rebase verification:
- main render path: hard clamp when fully bypassed, soft limiter otherwise;
- gapless transition tail follows the same bounding rule;
- out-of-order `setEqEnabled` / `setEqGains` responses and failures cannot overwrite newer state;
- reset cancels pending debounced gain commits;
- full CI + Packaging on the rebased head;
- manual EQ UI, audible bypass/DSP, clipping, and device-restart acceptance.

This rebased PR (or a focused replacement hotfix) should be the next audio fix landed on `main`.

**Comment by thedavidweng** [comment]
## Closing: All EQ fixes have been absorbed into main via #124

After comparing the latest #114 head against current `main`, **every EQ fix in this PR is already in main** — and main has advanced beyond it:

| Area | #114 | main | Status |
|------|------|------|--------|
| `eq.rs` | Per-channel coefficient computation | Per-band coefficient caching (#88 optimization) | main is **ahead** |
| `settings-store.ts` EQ guard | `eqMutationGeneration` counter | `OptimisticField` pattern with begin/confirm/reject | main is **ahead** (more sophisticated) |
| `output.rs` bypass safety | Hard clamp when bypassed | Hard clamp + soft limiter gating (#124) | main is **ahead** |
| `output.rs` gapless | Missing | Full gapless swap + FadingIn/FadingOut guards (#103, #124) | main is **ahead** |
| `settings.rs` | Missing `library_sort_mode`, `theme_preference` | Both present | main is **ahead** |

### What was compared
- `eq.rs`: main has 24 lines #114 doesn't (perf optimization); #114 has 10 lines main doesn't (older per-channel loop)
- `settings-store.ts`: main uses `OptimisticField<T>` with generation tracking; #114 uses simpler `eqMutationGeneration`
- `output.rs`: main has PeakRing, OutputFormatState, gapless swap, FadingIn exception; #114 has none of these
- `config.rs` / `settings.rs`: main has `LibrarySortMode`, `ThemePreference`; #114 doesn't

### Conclusion
`#124` ("restore mainline EQ and gapless safety") superseded this PR by incorporating all EQ fixes into main using a more robust `OptimisticField` pattern. The remaining #114 commits only re-add EQ infrastructure that already exists in main.

Closing as obsolete. If any specific fix is believed missing, please open a new PR rebased on current main with only the missing change.
